pkg/storage/ingest/pusher.go

```diff
@@ -212,62 +224,117 @@ type parallelStoragePusher struct {
 	}
 }
 
 // PushToStorage implements the PusherCloser interface.
-func (c parallelStoragePusher) PushToStorage(ctx context.Context, wr *mimirpb.WriteRequest) error {
+func (c *parallelStoragePusher) PushToStorage(ctx context.Context, wr *mimirpb.WriteRequest) error {
 	userID, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		level.Error(c.logger).Log("msg", "failed to extract tenant ID from context", "err", err)
 	}
 
 	shards := c.shardsFor(userID, wr.Source)
-	return shards.ShardWriteRequest(ctx, wr)
+	return shards.PushToStorage(ctx, wr)
 }
 
 // Close implements the PusherCloser interface.
-func (c parallelStoragePusher) Close() []error {
+func (c *parallelStoragePusher) Close() []error {
 	var errs multierror.MultiError
 	for _, p := range c.pushers {
-		errs.Add(p.Stop())
+		errs = append(errs, p.Close()...)
 	}
+	c.metrics.shardsPerPush.Observe(float64(c.numActiveShards))
+	c.metrics.pushersPerPush.Observe(float64(len(c.pushers)))
 	clear(c.pushers)
 	return errs
 }
 
 // shardsFor returns the parallelStorageShards for the given userID. Once created the same shards are re-used for the same userID.
 // We create a shard for each tenantID to parallelize the writes.
-func (c parallelStoragePusher) shardsFor(userID string, requestSource mimirpb.WriteRequest_SourceEnum) *parallelStorageShards {
+func (c *parallelStoragePusher) shardsFor(userID string, requestSource mimirpb.WriteRequest_SourceEnum) PusherCloser {
 	// Construct the string inline so that it doesn't escape to the heap. Go doesn't escape strings that are used to only look up map keys.
 	// We can use "|" because that cannot be part of a tenantID in Mimir.
 	if p := c.pushers[userID+"|"+requestSource.String()]; p != nil {
 		return p
 	}
 	// Use the same hashing function that's used for stripes in the TSDB. That way we make use of the low-contention property of stripes.
 	hashLabels := labels.Labels.Hash
-	p := newParallelStorageShards(c.metrics, c.errorHandler, c.numShards, c.batchSize, batchingQueueCapacity, c.upstreamPusher, hashLabels)
+
+	idealShards := c.IdealShardsFor(userID)
+	var p PusherCloser
+	if idealShards <= 1 {
+		// If we're going to push only one shard, then we can use the sequential pusher.
+		// This means that pushes will now be synchronous.
+		// The idea is that if we don't see a reason to parallelize,
+		// then the pushes to this pusher are likely small in absolute terms and speeding them up will have marginal gains.
+		// So we choose the lower overhead and simpler sequential pusher.
+		p = newSequentialStoragePusherWithErrorHandler(c.metrics, c.upstreamPusher, c.errorHandler)
+	} else {
+		p = newParallelStorageShards(c.metrics, c.errorHandler, idealShards, c.batchSize, c.queueCapacity, c.upstreamPusher, hashLabels)
+	}
 	c.pushers[userID+"|"+requestSource.String()] = p
 	return p
 }
 
+// IdealShardsFor returns the number of shards that should be used for the given userID.
+func (c *parallelStoragePusher) IdealShardsFor(userID string) int {
```

Comment:
[nit] I don't think we need to expose this funciton. `idealShardsFor()` should be enough.